pageserver/src/walingest.rs

```diff
@@ -343,7 +343,25 @@ impl WalIngest {
                         xlog_checkpoint.oldestActiveXid,
                         self.checkpoint.oldestActiveXid
                     );
-                    self.checkpoint.oldestActiveXid = xlog_checkpoint.oldestActiveXid;
+
+                    // A shutdown checkpoint has `oldestActiveXid == InvalidTransactionid`,
+                    // because at shutdown, all in-progress transactions will implicitly
+                    // end. Postgres startup code knows that, and allows hot standby to start
+                    // immediately from a shutdown checkpoint.
+                    //
+                    // In Neon, Postgres hot standby startup always behaves as if starting from
+                    // an online checkpoint. It needs a valid `oldestActiveXid` value, so
+                    // instead of overwriting self.checkpoint.oldestActiveXid with
+                    // InvalidTransactionid from the checkpoint WAL record, set it to `nextXid`
+                    // to indicate that all earlier transactions have finished. See also the
+                    // neon code changes in the InitWalRecovery() function.
+                    if xlog_checkpoint.oldestActiveXid == pg_constants::INVALID_TRANSACTION_ID
+                        && info == pg_constants::XLOG_CHECKPOINT_SHUTDOWN
+                    {
+                        self.checkpoint.oldestActiveXid = self.checkpoint.nextXid.value as u32;
+                    } else {
+                        self.checkpoint.oldestActiveXid = xlog_checkpoint.oldestActiveXid;
+                    }
```

Comment:
This is still confusing in the presence of prepared transactions. Normally, on an online checkpoint record, `oldestActiveXid` is always <= oldest prepared transaction. In other words, `oldestActiveXid` covers prepared transactions too. But here, it does not.

Reply:
In case of normal Postgres shutdown, checkpoint always contains `oldestActiveXid=0`, doesn't matter whether there are active prepared transactions or not.

This is why `StartupXLOG` is calling `PrescanPreparedTransactions` to correctly handle prepared transactions:

```
			if (wasShutdown)
				oldestActiveXID = PrescanPreparedTransactions(&xids, &nxids);
```

So I do not see any problems with replaced `oldestActiveXid=0' with `oldestActiveXid=nextXid' in this case.
Sorry, if I do not understand your concern.

Reply:
On PostgreSQL, `oldestActiveXid` on a checkpoint record means that there are no in-progress XIDs older than that. Including prepared transactions.

On Neon, `oldestActiveXid` on a checkpoint record means that there are no in-progress XIDs older than that. *Not* including prepared transactions.

I'm not sure what the implications of that are yet, but changing the meaning of a field like that is usually dangerous

Reply:
this is only different on the checkpoint record that's included in the control file. For the checkpoint records written to the WAL, `oldestActiveXid` means the same as in Postgres.

Reply:
Sorry, looks like I still missing something...
Vanilla Postgres always assign oldestActiveXid=0 in case of shutdown:
```
	/*
	 * For Hot Standby, derive the oldestActiveXid before we fix the redo
	 * pointer. This allows us to begin accumulating changes to assemble our
	 * starting snapshot of locks and transactions.
	 */
	if (!shutdown && XLogStandbyInfoActive())
		checkPoint.oldestActiveXid = GetOldestActiveTransactionId();
	else
		checkPoint.oldestActiveXid = InvalidTransactionId;
```
Doesn't matter wether there are prepared transactions or not.
And we can use `pg_controldata` to prove that `oldestActiveXid=0` after normal shutdown even if there were prepared transactions.

So I do not understand your statement that vanilla Postgres includes information about prepared transaction while Neon - doesn't.  I do not see here any difference with Vanilla.

Reply:
GetOldstActiveTransactonId() includes prepared transaction XIDs.

There are two cases in PostgreSQL:

wasShutdown==true: oldestActveXid is 0 in the checkpoint, and we use nextXid and PrescanPreparedTransactions to calculate the correct value.

wasShutdown==false: oldestActiveXid on the checkpoint is valid, _and it includes all prepared transaction XIDs_.


On Neon with this PR, when the pageserver sees a shutdown checkpoint, it sets oldestActiveXID to `nextXid`, which does _not_ include prepared transaction XIDs.

Reply:
But in Neon we also call `PrescanPreparedTransactions` in `RestoreRunningXactsFromClog`!
What is the difference?

Reply:
Well yeah I'm just trying to understand what the consequences are :-).

Found one issue at least: `StartupSUBTRANS(oldestActiveXID)` initialized pg_subtrans with a too new XID. That leads to "file not found" errors later if you need to look up older XIDs. I added a test case to this PR that demonstrates that.

Reply:
I think we should call PrescanPreparedTransactions earlier, in InitWalRecovery(). We currently have this there:
```

		if (!TransactionIdIsNormal(ControlFile->checkPointCopy.oldestActiveXid))
		{
			/*
			 * No checkpoint or running-xacts record was written, so use most
			 * conservative approximation for oldestActiveXid: oldest CLOG XID.
			 */
			ControlFile->checkPointCopy.oldestActiveXid = ControlFile->checkPointCopy.oldestXid;
		}
```
But that should never be necessary, now that we keep oldestActiveXid up-to-date in the pageserver. (Although maybe we need something to deal with old tenants that have had their WAL ingested with older pageserver versions before this PR was applied).

It still makes me a bit uneasy that the oldestActiveXid value in the control file's copy of the checkpoint is bogus, i.e. it doesn't include the prepared XIDs. We'd then  fix it up in InitWalRecovery(), so I guess it's fine, but it feels error-prone.

If you export a "full backup" to vanilla PostgreSQL, for example, would the bogus oldestActiveXID cause problems? I guess you cannot start up a vanillla Postgres server on a neon "fullbackup" export currently anywa.

Yet another approach would be to do the equivalent of PrescanPreparedTransactions in the pageserver, so that oldestActiveXID is always set correctly in the control file.  That seems pretty straightforward too: we have the code to list all pg_twophase entries in the pageserver already.

Reply:
> But that should never be necessary, now that we keep oldestActiveXid up-to-date in the pageserver. (

`oldestActiveXid` is advanced at PS when we replay checkpoint record, `oldestXid` - when we replay CLOG truncate record.  If we do not take `oldestXid` in account, we can try to access truncated CLOG segment.
So I think this check is still needed

Reply:
> Yet another approach would be to do the equivalent of PrescanPreparedTransactions in the pageserver, so that oldestActiveXID is always set correctly in the control file. That seems pretty straightforward too: we have the code to list all pg_twophase entries in the pageserver already.

I pushed a fix using that approach, seems the most straightforward

Reply:
> If you export a "full backup" to vanilla PostgreSQL, for example, would the bogus oldestActiveXID cause problems?

Does full backup enforce checkpoint which will contain valid `oldestActiverXid`?

Reply:
> > Yet another approach would be to do the equivalent of PrescanPreparedTransactions in the pageserver, so that oldestActiveXID is always set correctly in the control file. That seems pretty straightforward too: we have the code to list all pg_twophase entries in the pageserver already.
> 
> I pushed a fix using that approach, seems the most straightforward

Yes, it seems the simplest solution.
Should we then remove call of `PrescanPreparedTransactions` from `RestoreRunningXactsFromClog` in neon.c?